urlac47192b66_ggrandes_bouncer_tgz-pJ8-org_javastack_bouncer_ClusterTestJ8/src/org/javastack/bouncer/MuxServer.java

```diff
@@ -0,0 +1,566 @@
+package org.javastack.bouncer;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.net.SocketTimeoutException;
+import java.security.GeneralSecurityException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+
+import javax.net.ssl.SSLSocket;
+
+// ============================================ Mux Server
+
+// MuxServer (MUX=IN) Local=MUX, Remote=RAW
+class MuxServer {
+	final MuxServerMessageRouter router = new MuxServerMessageRouter();
+	final HashMap<Integer, MuxServerRemote> mapRemotes = new HashMap<Integer, MuxServerRemote>();
+	final ServerContext context;
+
+	MuxServerListenLocal localListen;
+	ArrayList<MuxServerListenRemote> remoteListen = new ArrayList<MuxServerListenRemote>(1);
+	MuxServerLocal local = null;
+
+	final InboundAddress left;
+	final Map<Integer, InboundAddress> right;
+
+	MuxServer(final ServerContext context, final InboundAddress left, final InboundAddress right) {
+		this.context = context;
+		this.left = left;
+		this.right = (right == null ? new HashMap<Integer, InboundAddress>() : //
+				Collections.singletonMap(Integer.valueOf(0), right));
+	}
+
+	void addRight(final InboundAddress right) {
+		this.right.put(right.getOpts().getTunID(), right);
+	}
+
+	void listenLocal() throws IOException { // Entry Point
+		localListen = new MuxServerListenLocal(left); // Local is MUX
+		context.addReloadableAwaiter(localListen);
+		context.submitTask(localListen, "MuxInListenLeft[" + left + "]", ClientId.newId());
+	}
+
+	void listenRemote() throws IOException {
+		for (InboundAddress right : this.right.values()) {
+			try {
+				final MuxServerListenRemote remoteListen = new MuxServerListenRemote(right); // Remote is RAW
+				this.remoteListen.add(remoteListen);
+				context.addReloadableAwaiter(remoteListen);
+				context.submitTask(remoteListen, "MuxInListenRight[" + left + "|" + right + "]",
+						ClientId.newId());
+			} catch (IOException e) {
+				Log.error(this.getClass().getSimpleName() + "::listenRemote " + e.toString(), e);
+			}
+		}
+	}
+
+	void closeRemote() {
+		for (final MuxServerListenRemote remoteListen : this.remoteListen) {
+			remoteListen.setShutdown();
+		}
+	}
+
+	void closeRemote(final int id) {
+		// Send FIN
+		try {
+			final MuxPacket mux = context.allocateMuxPacket();
+			mux.fin(id);
+			local.sendLocal(mux);
+			context.releaseMuxPacket(mux);
+		} catch (Exception ign) {
+		}
+		//
+		synchronized (mapRemotes) {
+			final MuxServerRemote remote = mapRemotes.remove(id);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *remote*. Execution paths that do not contain closure statements were detected.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).